util/resolver/pool.go

```diff
@@ -115,8 +135,33 @@ func newResolver(hosts docker.RegistryHosts, handler *authHandlerNS, sm *session
 	return r
 }
 
+func newOCILayoutResolver(layoutPath string, sm *session.Manager, g session.Group) *OCILayoutResolver {
+	fileStore, err := local.NewStore(layoutPath)
```

Comment:
No idea what this is supposed to be. I can't understand what is passed in here and the PR does not seem to even build. You need to pull in the blobs from the client side through a session API that is exposed by the client. Not allowed to read any files from host disk based on user input. Also `local.NewStore` is not OCI layout but containerd contentstore boltdb layout.

Reply:
> the PR does not seem to even build

I didn't expect it to. This is just for design review, to see that we are moving in the right direction and take course corrections.

> No idea what this is supposed to be. I can't understand what is passed in here

I largely copied what the existing resolver from the registry does. Instead of a ref, I passed where the expected oci-layout directory is.

> Also local.NewStore is not OCI layout but containerd contentstore boltdb layout.

I know. But it happens to be oci-layout compliant except for the stuff that goes in boltdb (image to hash mapping, etc.). This way, we inherit the reading of hashes from containerd's library. I could write it from scratch, but why bother? It does all of the reading, validation, etc. And buildkit already imports containerd packages.

> You need to pull in the blobs from the client side through a session API that is exposed by the client. Not allowed to read any files from host disk based on user input.

This is precisely what I punted on (see the PR opening comment). I wasn't sure how we do that, as buildkit runs potentially on a different host than the client.

How does `local` source do that? Does the client recognize the option, mount the directory, and stream it to buildkitd? Can you share a link to consuming that session API?

Reply:
Client exposes service on session with `s.Allow` https://github.com/moby/buildkit/blob/master/client/solve.go#L127 . Daemon side gets the caller for it https://github.com/moby/buildkit/blob/master/source/local/local.go#L98-L101 .

For the contentstore this already exists for local cache import https://github.com/moby/buildkit/blob/master/client/solve.go#L161 so basically you need to attach your contentstore to this same map. Then pass the sessionID and digest with the LLB definition so the daemon side can provide it as input when accessing the session service.

Reply:
> when the client sets up grpc to the daemon, the internal structure that represents that communication is a session

When you invoke a typical build, client does 3 internal requests separately.
- solve/build requests with your build definition. This blocks until build is complete.
- request to receive progress events from a build.
- a session requests. session is a tunnel that can be used by the builder to pull in allowed resources from the client. eg. files from build context, authentication tokens etc.

> anything daemon side

Correct. You can get a caller from the session group that is passed to the ops methods. Session caller can be used to create a grpc client to access the specific service client exposed.

> why do we have both [this s.Allow()](https://github.com/moby/buildkit/blob/master/client/solve.go#L161) and [this one](https://github.com/moby/buildkit/blob/master/client/solve.go#L127)?

There are many different services exposed with a session. The ones you linked are for transferring local directories for build context(s) (used with `--local` in `buildctl`) and cache-from with `type=local`. The second one is contentstore interface while first is filesync API for transferring directories. There are many more, access to auth tokens, build secrets, ssh agents, tar upload etc. https://github.com/moby/buildkit/tree/master/session

> somewhere, it must parse the options (e.g. --build-context)

`build-context` is a Dockerfile feature passed with `--opt context:name=value`. Don't bother with that on the first PR. For now your goal is to add this feature to LLB with tests. Then it can be connected with Dockerfile later.

On client side your goal is to extend current `llb.Image`, `llb.Git`, `llb.Local` etc. with something like `llb.OCIArchive(contentStoreID, digest)`.

> Does the filesync just allow daemon-side

Filesync allows pulling directories from client to daemon(and incremental updates of them) https://pkg.go.dev/github.com/moby/buildkit@v0.10.2/session/filesync#FileSyncServer . For this you need https://pkg.go.dev/github.com/moby/buildkit@v0.10.2/session/content https://pkg.go.dev/github.com/containerd/containerd@v1.6.3/content/proxy#NewContentStore https://pkg.go.dev/github.com/containerd/containerd@v1.6.3/api/services/content/v1#ContentServer

You will not make a full copy of the OCI layout directory and then start to parse it. There is already an API that will directly pull the blobs over grpc the same way as they are pulled from the registry.

Reply:
> You will not make a full copy of the OCI layout directory and then start to parse it.

Hell no. We should know exactly what file we are after, and should be able to pull that. But I think you are saying that there already is an API built to turn the remote end into a containerd `content.Store`, so we do not even need to do that? This:

> There is already an API that will directly pull the blobs over grpc the same way as they are pulled from the registry.

Are you saying that I can use [proxy.NewContentStore](https://pkg.go.dev/github.com/containerd/containerd@v1.6.3/content/proxy#NewContentStore) to get a `content.Store` which I can use to just pull blobs as needed, thus greatly simplifying the daemon side?

And that something in session already satisfies the [contentapi.ContentClient](https://pkg.go.dev/github.com/containerd/containerd@v1.6.3/api/services/content/v1#ContentClient) which I need to pass to [NewContentStore](https://pkg.go.dev/github.com/containerd/containerd@v1.6.3/content/proxy#NewContentStore)?

Or are you saying that is the way to go, but we do not (yet) have the construct within the session that satisfies [contentapi.ContentClient](https://pkg.go.dev/github.com/containerd/containerd@v1.6.3/api/services/content/v1#ContentClient), and thus we need to create it?

Reply:
And now I think I see that [session/content.NewCallerStore](https://pkg.go.dev/github.com/moby/buildkit@v0.10.2/session/content#NewCallerStore) seems to take a `session.Caller` and return a `content.Store`.

Do I call `NewCallerStore(caller)` and get a `content.Store` out of it, that I can use to read blobs? If so, what do I need `FileSync` for?

Reply:
> If so, what do I need FileSync for?

You don't. I'm not sure why you would think otherwise.

Reply:
> You don't.

Good.

OK, so I can pass a `session.Caller` to `session/content.NewCallerStore` and get a `content.Store`. From that I can read any blob.

* where does [proxy.NewContentStore()](https://pkg.go.dev/github.com/containerd/containerd@v1.6.3/content/proxy#NewContentStore) fit into all of this?
* what do I need to pass to `s.Allow()` client-side to enable it to serve up the files? It looks like it has something to do with [RegisterContentServer](https://pkg.go.dev/github.com/containerd/containerd@v1.6.3/api/services/content/v1#RegisterContentServer), but I am not at all sure: a) that is correct; b) how that plays into `s.Allow()`; c) what I would pass to as the [ContentServer](https://pkg.go.dev/github.com/containerd/containerd@v1.6.3/api/services/content/v1#ContentServer) to `RegisterContentServer`

And all of this may be barking up the wrong tree?

Reply:
> where does [proxy.NewContentStore()](https://pkg.go.dev/github.com/containerd/containerd@v1.6.3/content/proxy#NewContentStore) fit into all of this?

It is called inside `session/content.NewCallerStore`

> what do I need to pass to s.Allow()

You need to make sure https://github.com/moby/buildkit/blob/master/client/solve.go#L161 gets called and that this map contains the contentstore that provides access to the OCI archive(s) you are exposing.

Reply:
> It is called inside session/content.NewCallerStore

Lower-layer, so I can ignore it at this level. Works for me.

> You need to make sure https://github.com/moby/buildkit/blob/master/client/solve.go#L161 gets called and that this map contains the contentstore that provides access to the OCI archive(s) you are exposing.

He makes it sound so simple. I wish!

> this map contains the contentstore that provides access to the OCI archive(s) you are exposing.

Do I just need to create a `content.Store` that accesses the local filesystem path and pass it?

Reply:
Trying to figure out the client-side. I don't see documented anywhere how the `docker buildx` concept of `build-context` translates into something that buildkit understands.

The [s.Allow(...)](https://github.com/moby/buildkit/blob/master/client/solve.go#L161) uses `cacheOpt.contentStores`. As far as I can tell, those are loaded via [parseCacheOptions](https://github.com/moby/buildkit/blob/master/client/solve.go#L392-L505), which only include reading cache imports and cache exports.

Where does the "contexts" (or aliasing) take place? Given that I could not find it anywhere in buildkit, I suspect that it is a "docker buildx" thing which is "translated" into some buildkit concept related to caches?

I tested that hypothesis by looking for where it might be importing local and git. I found local in cache import (which does update `contentStores`), but not git, so maybe I am barking up the wrong tree.

To translate to real questions:

* what does the `--build-context foo=oci-layout://...` get translated to in terms of buildkit options? What does it look like?
* where does that get parsed, such that I can extract the path and pass the local content store into `contentStores`?
* where does the image name side of it (e.g. `bar=./path/to/dir`) get parsed, i.e. the `bar` part? Half of it is loading `path/to/dir` into the content stores, but the other half is telling buildkit, "when looking for `bar`, this is the content store you use".
* how does it translate the image name into the root manifest? An OCI layout can have many manifests.

Reply:
> Do I just need to create a content.Store that accesses the local filesystem path and pass it?

yes

>  I don't see documented anywhere how the docker buildx concept of build-context translates into something that buildkit understands.

It's in Dockerfile frontend. Read https://github.com/moby/buildkit/pull/2827#discussion_r862207587

> what does the --build-context foo=oci-layout://... get translated to in terms of buildkit options? What does it look like?

Client converts the local directory to contentstore(and finds the target digest). Add `OCIArchives map[string]content.Store` to https://pkg.go.dev/github.com/moby/buildkit@v0.10.2/client#SolveOpt . In `FrontendOpts/FrontendAttr` buildx would set set `context:foo=oci-layout://contentstoreID/image-digest`. Handler for the `FrontendOpts` in Dockerfile should be in a later PR(and `--build-context` is in buildx). Only LLB support for now.

> where does that get parsed, such that I can extract the path and pass the local content store into contentStores?

Only buildx implements this flag. If we want `buildctl` support then the flag for that would be something like `--oci-archive digest=./path/to/store` similar how `buildctl` exposes `--local`. But for me just Go LLB library support for this is ok for now.

> where does the image name side of it

No idea what you mean in here. Maybe answered above.

> how does it translate the image name into the root manifest? An OCI layout can have many manifests.

This is for the client side to decide when it parses `--build-context` value. For the daemon, the only inputs are `contentStoreID` and `digest`. Multiple digests could come from same contentstore.

Reply:
Going to give this a shot @tonistiigi ; tell me what I got right and wrong.

* the daemon is responsible for the build itself. It doesn't understand dockerfile or other frontends; it gets an LLB via a [SolveRequest](https://github.com/moby/buildkit/blob/master/api/services/control/control.proto#L54-L65) and processes that.
* the client is responsible for parsing the frontend (e.g. Dockerfile) into an LLB, creating that `SolveRequest` and sending it to the daemon (as well as awaiting status updates and opening the channel to send requested info by the daemon).

The process looks like:

1. client starts a build with Dockerfile (or other frontend) and context(s) and options
2. client (built-in or via some other frontend processor) processes the frontend, getting to an LLB, which includes some sources
3. sends the `SolveRequest` to the daemon, which includes each source's, well, "source"

In practical terms, this means that if I have the following:

```dockerfile
FROM golang:latest as builder
RUN something
FROM alpine:3.15
COPY --from=builder /else /here
```

The determining of `builder` as based on the image `golang:latest` happens on the client-end; buildkitd doesn't know or understand any of that. The _client_ using a frontend processor determines "I need to resolve `golang:latest` and `alpine:3.15`", and the _client_ figures out how to resolve each of those. It just passes each of those as a source. In the case of an image, that source will be a registry reference; in the case of local, it will be give it some reference that says, "go request this reference thing from the channel I opened with you."

Translating all of that to OCI, the _client_:
1. calls the frontend processor (unchanged)
2. sees that one of the sources is `golang:latest` (unchanged)
3. knows (based on its CLI flags or whatever) that `golang:latest` should be converted to a hash in a local `content.Store` (new)
4. sends the LLB to the daemon, along with the source for that particular element as "go request this reference thing from the channel I opened with you."

For that to work:

* the daemon needs to know that it has a new kind of source and how to get it (similar to local: it just reaches back on the channel to the client; similar to registry: everything is hashes and manifests)
* the client needs to know, when its frontend processor has an image that matches some indication (i.e. new context given), that it should find the local directory, open a local `content.Store` and make it available via the channel

Everything we did in this PR to date was about the first half of that: getting the daemon to recognize a new source reference that the client will pass it in its LLB; everything in the follow-on here is about getting the client to know how to find those source and create the correct reference to pass to the daemon.

If that is correct, let me know and I will try to interpret your above comments in that light.

Reply:
Clients sends a Solve or Build request to build using a Dockerfile frontend. Dockerfile frontend runs inside a container, parses the Dockerfile, generates LLB and issues LLB Solve requests.

Reply:
Buildkitd runs in one container; frontend can be in another. So client issues solve requests to frontend container, gets response as a whole bunch of LLB Solve requests, and then passes those to buildkitd?

Reply:
No, frontend issues the LLB Solve requests, not a client(when building a Dockerfile).